### PR TITLE
Upgrade React, react-redux, connected-react-router and testing-library/react to the latest versions

### LINKED ISSUE #11638: Update React monorepo
We should update to the latest React. Renovate has a PR open at https://github.com/mozilla/addons-frontend/pull/11401, but this may be a large task, so it might need this as a tracking issue.

### LINKED ISSUE #11639: Update react-router and connected-react-router
We should update to the latest versions of react-router and connected-react-router. Renovate has PRs open at https://github.com/mozilla/addons-frontend/pull/10866 and https://github.com/mozilla/addons-frontend/pull/10742, but this may be a large task, so it might need this as a tracking issue.

### LINKED ISSUE #11641: Update react-redux
We should update to the latest react-redux. Renovate has a PR open at https://github.com/mozilla/addons-frontend/pull/11400, but this may be a large task, so it might need this as a tracking issue.

### LINKED ISSUE #11644: Update to React 18
This is a spin-off from https://github.com/mozilla/addons-frontend/issues/11638. Once we are on React 17, we can attempt to update to React 18.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #11639 
Fixes #11644 
Fixes #11641 
Fixes #11638 

This has React 18, Redux 8, connected-react-router 6 and react-testing-library 13.

It works locally, but there are lots of failing tests. I am working my way through them.

For QA:

This updates our major dependencies. It should not result in any behavioural changes, but it is a big change, so the site should be tested thoroughly. One thing in particular to test, because there were some oddities with changes we had to make to tests around this, is the feature that does a redirect when a user visits a collection page using a userName instead of a userId, as well as when a user visits a collection page using a slug that has a different case than the actual collection slug.

### CODE REVIEW

**Comment by diox** [comment]
Pasting this here to look at at some point:

```
warning " > react-nested-status@0.2.1" has incorrect peer dependency "react@^16.0.0".
warning "react-nested-status > react-side-effect@1.2.0" has incorrect peer dependency "react@^0.13.0 || ^0.14.0 || ^15.0.0 || ^16.0.0".
warning " > react-onclickoutside@6.12.1" has incorrect peer dependency "react@^15.5.x || ^16.x || ^17.x".
warning " > react-onclickoutside@6.12.1" has incorrect peer dependency "react-dom@^15.5.x || ^16.x || ^17.x".
warning " > react-super-responsive-table@5.2.0" has incorrect peer dependency "react@^16.9.0 || ^17.0.0".
warning " > react-super-responsive-table@5.2.0" has incorrect peer dependency "react-dom@^16.9.0 || ^17.0.0".
warning " > react-textarea-autosize@8.3.3" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
warning "react-textarea-autosize > use-composed-ref@1.1.0" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
warning "react-textarea-autosize > use-latest@1.2.0" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
warning "react-textarea-autosize > use-latest > use-isomorphic-layout-effect@1.1.1" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react@^15.0.0 || ^16.0.0 || ^17.0.0 ".
warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react-dom@^15.0.0 || ^16.0.0 || ^17.0.0 ".
warning " > react-test-renderer@16.14.0" has incorrect peer dependency "react@^16.14.0".
```

**Comment by bobsilverberg** [comment]
> Pasting this here to look at at some point:
> 
> ```
> warning " > react-nested-status@0.2.1" has incorrect peer dependency "react@^16.0.0".
> warning "react-nested-status > react-side-effect@1.2.0" has incorrect peer dependency "react@^0.13.0 || ^0.14.0 || ^15.0.0 || ^16.0.0".
> warning " > react-onclickoutside@6.12.1" has incorrect peer dependency "react@^15.5.x || ^16.x || ^17.x".
> warning " > react-onclickoutside@6.12.1" has incorrect peer dependency "react-dom@^15.5.x || ^16.x || ^17.x".
> warning " > react-super-responsive-table@5.2.0" has incorrect peer dependency "react@^16.9.0 || ^17.0.0".
> warning " > react-super-responsive-table@5.2.0" has incorrect peer dependency "react-dom@^16.9.0 || ^17.0.0".
> warning " > react-textarea-autosize@8.3.3" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> warning "react-textarea-autosize > use-composed-ref@1.1.0" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> warning "react-textarea-autosize > use-latest@1.2.0" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> warning "react-textarea-autosize > use-latest > use-isomorphic-layout-effect@1.1.1" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react@^15.0.0 || ^16.0.0 || ^17.0.0 ".
> warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react-dom@^15.0.0 || ^16.0.0 || ^17.0.0 ".
> warning " > react-test-renderer@16.14.0" has incorrect peer dependency "react@^16.14.0".
> ```

Thanks @diox. I noticed these as well, and it looks like most of them do have minor updates that should fix these warnings, but I decided to leave these for a future patch. I guess we should open an issue for that update.

**Comment by bobsilverberg** [comment]
> > Pasting this here to look at at some point:
> > ```
> > warning " > react-nested-status@0.2.1" has incorrect peer dependency "react@^16.0.0".
> > warning "react-nested-status > react-side-effect@1.2.0" has incorrect peer dependency "react@^0.13.0 || ^0.14.0 || ^15.0.0 || ^16.0.0".
> > warning " > react-onclickoutside@6.12.1" has incorrect peer dependency "react@^15.5.x || ^16.x || ^17.x".
> > warning " > react-onclickoutside@6.12.1" has incorrect peer dependency "react-dom@^15.5.x || ^16.x || ^17.x".
> > warning " > react-super-responsive-table@5.2.0" has incorrect peer dependency "react@^16.9.0 || ^17.0.0".
> > warning " > react-super-responsive-table@5.2.0" has incorrect peer dependency "react-dom@^16.9.0 || ^17.0.0".
> > warning " > react-textarea-autosize@8.3.3" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> > warning "react-textarea-autosize > use-composed-ref@1.1.0" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> > warning "react-textarea-autosize > use-latest@1.2.0" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> > warning "react-textarea-autosize > use-latest > use-isomorphic-layout-effect@1.1.1" has incorrect peer dependency "react@^16.8.0 || ^17.0.0".
> > warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react@^15.0.0 || ^16.0.0 || ^17.0.0 ".
> > warning " > react-hot-loader@4.13.0" has incorrect peer dependency "react-dom@^15.0.0 || ^16.0.0 || ^17.0.0 ".
> > warning " > react-test-renderer@16.14.0" has incorrect peer dependency "react@^16.14.0".
> > ```
> 
> Thanks @diox. I noticed these as well, and it looks like most of them do have minor updates that should fix these warnings, but I decided to leave these for a future patch. I guess we should open an issue for that update.

I have included a few more updates that address some of these peer dependency warnings. Unfortunately a few of them have not been addressed in the libraries themselves so the warnings for those remain.

File: webpack.dev.config.babel.js [hunk 1]
```diff
@@ -59,14 +59,14 @@ export default {
...
-      'react-dom/umd/react-dom.development.js',
-    ),
+    // new webpack.NormalModuleReplacementPlugin(
+    //   /^react$/,
+    //   'react/umd/react.development.js',
+    // ),
+    // new webpack.NormalModuleReplacementPlugin(
+    //   /^react-dom$/,
+    //   'react-dom/umd/react-dom.development.js',
+    // ),
```
Comment (eviljeff) [inline webpack.dev.config.babel.js:69]: Is this supposed to be commented out?  (should it be deleted instead?)

Comment (bobsilverberg) [reply webpack.dev.config.babel.js:69]: Well... it's something I had to remove, and I was unable to find a replacement for it. I don't really want to just delete it, so I guess I should comment it out with a descriptive comment and open an issue to see if we can find a replacement, or determine if it's actually needed. I'll do that.

**Review by eviljeff — APPROVED** [review]

**Pull request merged**